packages/ws/src/ws/WebSocketShard.ts

```diff
@@ -234,18 +244,58 @@ export class WebSocketShard extends AsyncEventEmitter<WebSocketShardEventsMap> {
 		}
 	}
 
-	private async waitForEvent(event: WebSocketShardEvents, timeoutDuration?: number | null) {
-		this.debug([`Waiting for event ${event} for ${timeoutDuration ? `${timeoutDuration}ms` : 'indefinitely'}`]);
+	private async waitForEvent(event: WebSocketShardEvents, timeoutDuration?: number | null): Promise<void> {
+		this.debug([`Waiting for event ${event} ${timeoutDuration ? `for ${timeoutDuration}ms` : 'indefinitely'}`]);
 		const controller = new AbortController();
 		const timeout = timeoutDuration ? setTimeout(() => controller.abort(), timeoutDuration).unref() : null;
 		if (timeout) {
 			this.timeouts.set(event, timeout);
 		}
 
-		await once(this, event, { signal: controller.signal });
-		if (timeout) {
-			clearTimeout(timeout);
-			this.timeouts.delete(event);
+		await once(this, event, { signal: controller.signal }).finally(() => {
+			if (timeout) {
+				clearTimeout(timeout);
+				this.timeouts.delete(event);
+			}
+		});
+	}
+
+	/**
+	 * Does special error handling for waitForEvent calls, depending on the current state of the connection lifecycle
+	 * (i.e. whether or not the original connect() call has resolved or if the user has an error listener)
+	 * Returns null on success, or an error (unknown type) on failure
+	 */
+	private async bubbleWaitForEventError(promise: Promise<unknown>): Promise<unknown> {
+		try {
+			await promise;
+			return null;
+		} catch (error) {
+			// Any error that isn't an abort error would have been caused by us emitting an error event in the first place
+			// See https://nodejs.org/api/events.html#eventsonceemitter-name-options for `once()` behavior
+			if (error instanceof Error && error.name === 'AbortError') {
+				this.emit(WebSocketShardEvents.Error, { error });
+			}
+
+			// As stated previously, any other error would have been caused by us emitting the error event, which looks
+			// like { error: unknown }
+			// eslint-disable-next-line no-ex-assign
+			error = (error as { error: unknown }).error;
+
+			// If the user has no handling on their end (error event) simply throw.
+			// We also want to throw if we're still in the initial `connect()` call, since that's the only time
+			// the user can catch the error "normally"
+			if (this.listenerCount(WebSocketShardEvents.Error) === 0 || !this.initialConnectResolved) {
+				throw error;
+			}
+
+			// If the error is handled, we can just try to reconnect
+			await this.destroy({
+				code: CloseCodes.Normal,
+				reason: 'Something timed out',
+				recover: WebSocketShardDestroyRecovery.Reconnect,
+			});
+
+			return null;
```

Comment:
shouldn't this return something other than `null` to prevent the original connect method from continuing after the destroy method re-calls it, potentially identifying/resuming twice?

Reply:
Yes. I think ultimately I'll opt for a different signaling system - probably a result type of sorts with a bool prop on it.